Articles pagination

tyndria opened the pull request:
Add pagination for main page articles

tyndria committed:
Add api constant and redux-related logic for articles with pagination request; Start integrating pagination to the main page

tyndria committed:
Merge branch 'master' into articles-pagination

# Conflicts:
#	pages/index.js
#	redux/ducks/articles.js

tyndria committed:
Add function for splitting articles into rows

tyndria committed:
Merge master into articles-pagination branch. Fix conflicts

tyndria committed:
Refactor splitting articles row

tyndria committed:
Merge branch 'master' into articles-pagination

# Conflicts:
#	redux/ducks/articles.js
#	utils/customPropTypes.js
#	utils/getters.js

Drapegnik commented on constants/api.js:
```diff
@@ -9,6 +9,7 @@ export default {
   },
   articles: {
     getAll: `${API_URL}/articles`,
+    getChunk: (page, pageSize) => `${API_URL}/articles?page=${page}&pageSize=${pageSize}`,
```
* you should modify `getAll`, not create new endpoint
* `page=${page}` what about to use some default solutions (like [`qs`](https://github.com/ljharb/qs) for query params?

tyndria replied:
@Drapegnik the library is nice, agree, but does it make sense to connect it for a such simple case?

Drapegnik replied:
It's already used

tyndria replied:
hmm, I see it in package.json, but I can't find where you are using it

Drapegnik commented on pages/index.js:
```diff
@@ -36,38 +41,53 @@ class HomePage extends Component {
     }).isRequired,
     articles: ArticlesArray.isRequired,
     diary: DiaryShape.isRequired,
+    pagination: PropTypes.oneOfType([PropTypes.bool, PaginationShape]).isRequired,
```
`bool`?

tyndria replied:
It will be ``false`` when there are no more pages. not so obviously, agree(

Drapegnik commented on pages/index.js:
```diff
@@ -36,38 +41,53 @@ class HomePage extends Component {
+    const { articles, error, diary, pagination, getByDay, getChunk, url } = this.props;
+
+    const articlesRows = getArticlesRows(articles, ROW_SIZE);
+    const firstPageRows = { first: articlesRows[0], second: articlesRows[1] };
```
```js

Drapegnik replied:
```js
const [first, second] = aarticlesRows
```

Drapegnik commented on pages/index.js:
```diff
@@ -36,38 +41,53 @@ class HomePage extends Component {
+              )}
+            />
+
+            {articles.length > PAGE_SIZE &&
```
Please extract all this logic to component `Pagination`

Drapegnik replied:
Possible api:
```jsx
  <Pagination
    page={}
	pageSize={}
  />
```

Drapegnik commented on redux/ducks/articles.js:
```diff
@@ -36,13 +38,21 @@ const currentReducer = defaultReducer((state, { payload }) => ({
     })),
+    [FETCH_CHUNK]: defaultReducer((state, { payload: { data, next } }) => ({
+      ...state,
+      data: next.page === FIRST_PAGE_NUMBER + 1 ? data : [...state.data, ...data],
```
what about init data with `[]` & omit this ternary?

tyndria replied:
@Drapegnik , this action is dispatched when you are navigating across different tabs or come to the main page through Back button. In a such case the store isn't reloaded.
I could try to reload it as you do here https://github.com/babajka/babajka-frontend/blob/master/routes.js#L36 or leave code as it is

Drapegnik commented on utils/getters.js:
```diff
@@ -47,3 +47,11 @@ export const getLocalizedArticle = (article, lang) => {
 export const getLocalizedArticles = (articles, lang) =>
   articles.map(article => getLocalizedArticle(article, lang));
+
+export const getArticlesRows = (articles, rowSize) => {
```
check out [`lodash@chunk`](https://lodash.com/docs/4.17.4#chunk)

tyndria replied:
thanks! I will definitely use it

tyndria committed:
Fixes:
1. Refactor function for getting articles rows
2. Use one endpoint for articles fetch
3. Refactor splitting articles rows

Drapegnik commented on redux/ducks/articles.js:
```diff
@@ -69,11 +69,11 @@ export default createReducer(
+    payload: request.fetch(api.articles.getAll()),
   }),
-  fetchChunk: (page, pageSize) => ({
+  fetchChunk: (page = 0, pageSize = 8) => ({
```
use constants

Drapegnik commented on constants/api.js:
```diff
@@ -8,8 +8,10 @@ export default {
-    getChunk: (page, pageSize) => `${API_URL}/articles?page=${page}&pageSize=${pageSize}`,
+    getAll: (page, pageSize) => {
+      const baseUrl = `${API_URL}/articles`;
+      return page && pageSize ? `${baseUrl}?page=${page}&pageSize=${pageSize}` : baseUrl;
```
`page && pageSize ?` - you feel that this bicycle becomes more and more confused?

please use `qs`, or, if you want optimize bundle size, implement your own method `formatUrl(baseUrl, { page, pageSize })`

tyndria committed:
Refactor code:
1. Define constants
2. Use 'qs' to format url for getting articles chunk

tyndria committed:
Select just next page from the state instead of all pagination config; Do not render first/second rows if there are not enough articles

tyndria committed:
Merge branch 'master' into articles-pagination

tyndria committed:
Update styles

tyndria committed:
Add loader for Load more button

tyndria committed:
Merge branch 'master' into articles-pagination

# Conflicts:
#	utils/getters.js

Drapegnik committed:
task(articles): add some fixes
